File: service/src/main/java/no/nav/melosys/service/kontroll/feature/unntak/AnmodningUnntakKontrollService.java
package no.nav.melosys.service.kontroll.feature.unntak;

import java.util.Collection;
import java.util.Objects;

import no.nav.melosys.domain.Anmodningsperiode;
import no.nav.melosys.domain.Behandling;
import no.nav.melosys.service.avklartefakta.OppsummerteAvklarteFaktaService;
import no.nav.melosys.service.behandling.BehandlingService;
import no.nav.melosys.service.kontroll.feature.unntak.data.AnmodningUnntakKontrollData;
import no.nav.melosys.service.kontroll.feature.unntak.kontroll.AnmodningUnntakKontrollsett;
import no.nav.melosys.service.persondata.PersondataFasade;
import no.nav.melosys.service.unntak.AnmodningsperiodeService;
import no.nav.melosys.service.validering.Kontrollfeil;
import org.springframework.stereotype.Service;

@Service
public class AnmodningUnntakKontrollService {

    private final AnmodningsperiodeService anmodningsperiodeService;
    private final OppsummerteAvklarteFaktaService oppsummerteAvklarteFaktaService;
    private final BehandlingService behandlingService;
    private final PersondataFasade persondataFasade;

    public AnmodningUnntakKontrollService(AnmodningsperiodeService anmodningsperiodeService,
                                          OppsummerteAvklarteFaktaService oppsummerteAvklarteFaktaService, BehandlingService behandlingService,
                                          PersondataFasade persondataFasade) {
        this.anmodningsperiodeService = anmodningsperiodeService;
        this.oppsummerteAvklarteFaktaService = oppsummerteAvklarteFaktaService;
        this.behandlingService = behandlingService;
        this.persondataFasade = persondataFasade;
    }

    public Collection<Kontrollfeil> utførKontroller(long behandlingID) {
        return utførKontroller(
            behandlingService.hentBehandlingMedSaksopplysninger(behandlingID),
            anmodningsperiodeService.hentFørsteAnmodningsperiode(behandlingID)
        );
    }

    private Collection<Kontrollfeil> utførKontroller(
        Behandling behandling,
        Anmodningsperiode anmodningsperiode
    ) {
        final var persondata = persondataFasade.hentPerson(behandling.getFagsak().hentBrukersAktørID());
        final int antallArbeidsgivere = oppsummerteAvklarteFaktaService.hentAntallAvklarteVirksomheter(behandling);
        AnmodningUnntakKontrollData kontrollData = new AnmodningUnntakKontrollData(persondata,
            behandling.getMottatteOpplysninger().getMottatteOpplysningerData(),
            anmodningsperiode,
            antallArbeidsgivere
        );
        return AnmodningUnntakKontrollsett.hentRegler().stream()
            .map(f -> f.apply(kontrollData))
            .filter(Objects::nonNull)
            .toList();
    }
}

File: service/src/main/kotlin/no/nav/melosys/service/kontroll/feature/anmodningomunntak/AnmodningUnntakKontrollService.kt
package no.nav.melosys.service.kontroll.feature.anmodningomunntak

import no.nav.melosys.domain.Aktoer
import no.nav.melosys.domain.dokument.organisasjon.OrganisasjonDokument
import no.nav.melosys.domain.kodeverk.Fullmaktstype
import no.nav.melosys.domain.person.Persondata
import no.nav.melosys.service.avklartefakta.AvklarteVirksomheterService
import no.nav.melosys.service.behandling.BehandlingService
import no.nav.melosys.service.kontroll.feature.anmodningomunntak.data.AnmodningUnntakKontrollData
import no.nav.melosys.service.kontroll.feature.anmodningomunntak.kontroll.AnmodningUnntakKontrollsett
import no.nav.melosys.service.persondata.PersondataFasade
import no.nav.melosys.service.registeropplysninger.OrganisasjonOppslagService
import no.nav.melosys.service.unntak.AnmodningsperiodeService
import no.nav.melosys.service.validering.Kontrollfeil
import org.springframework.stereotype.Service


@Service
class AnmodningUnntakKontrollService(
    private val anmodningsperiodeService: AnmodningsperiodeService,
    private val avklarteVirksomheterService: AvklarteVirksomheterService,
    private val behandlingService: BehandlingService,
    private val persondataFasade: PersondataFasade,
    private val organisasjonOppslagService: OrganisasjonOppslagService
) {
    fun utførKontroller(behandlingID: Long): Collection<Kontrollfeil> {
        val behandling = behandlingService.hentBehandlingMedSaksopplysninger(behandlingID)
        val fullmektig = behandling.fagsak.finnFullmektig(Fullmaktstype.FULLMEKTIG_SØKNAD)

        val kontrollData = AnmodningUnntakKontrollData(
            persondata = persondataFasade.hentPerson(behandling.fagsak.hentBrukersAktørID()),
            mottatteOpplysningerData = behandling.mottatteOpplysninger.mottatteOpplysningerData,
            anmodningsperiode = anmodningsperiodeService.hentFørsteAnmodningsperiode(behandlingID),
            antallArbeidsgivere = avklarteVirksomheterService.hentAntallAvklarteVirksomheter(behandling),
            fullmektig = fullmektig,
            organisasjonDokumentTilFullmektig = hentOrganisasjonFullmektig(fullmektig),
            persondataTilFullmektig = hentPersondataFullmektig(fullmektig),
            medlemskapDokument = behandling.hentMedlemskapDokument(),
        )

        return AnmodningUnntakKontrollsett.hentRegler().mapNotNull { it.apply(kontrollData) }
    }

    private fun hentPersondataFullmektig(fullmektig: Aktoer?): Persondata? =
        if (fullmektig != null && fullmektig.erPerson()) persondataFasade.hentPerson(fullmektig.personIdent) else null

    private fun hentOrganisasjonFullmektig(fullmektig: Aktoer?): OrganisasjonDokument? =
        if (fullmektig != null && fullmektig.erOrganisasjon()) organisasjonOppslagService.hentOrganisasjon(fullmektig.orgnr) else null
}
